### PR TITLE
k8s 1.29 bump

### PR DESCRIPTION
**Pull request by kannon92**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Update kubernetes to 0.29 api changes.
#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
I see two groups of failures:
```
field.NotSupported(field.NewPath("spec", "tolerations").Index(2).Child("effect"), corev1.TaintEffect("not-valid"), nil),
```
Error is "cannot infer t" and these errors are in pkg/webhooks/resourceflavor_webhook_test.go and pkg/webhooks/workload_webhooks_test.go.

And the pkg/controller/job/pod has the following error while running `make test`.

```
--- FAIL: TestReconciler_ErrorFinalizingPod (0.00s)
    /home/kehannon/Work/kueue/pkg/controller/jobs/pod/reconciler.go:233: "level"=2 "msg"="Reconciling Job" "job"="ns/pod" "gvk"="/v1, Kind=Pod"
    /home/kehannon/Work/kueue/pkg/controller/jobs/pod/pod_controller_test.go:2213: Expected reconcile error (-want,+got):
          any(
        - 	e"connection refused: connection refused",
          )
    /home/kehannon/Work/kueue/pkg/controller/jobs/pod/reconciler.go:233: "level"=2 "msg"="Reconciling Job" "job"="ns/pod" "gvk"="/v1, Kind=Pod"
```

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -15,20 +15,20 @@ require (
...
+	k8s.io/component-base v0.29.1
+	k8s.io/component-helpers v0.29.1
+	k8s.io/klog/v2 v2.110.1
+	k8s.io/kube-openapi v0.0.0-20231010175941-2dd684a91f00
+	k8s.io/metrics v0.29.1
 	k8s.io/utils v0.0.0-20230726121419-3b25d923346b
-	sigs.k8s.io/controller-runtime v0.16.3
-	sigs.k8s.io/jobset v0.3.1
+	sigs.k8s.io/controller-runtime v0.17.0
+	sigs.k8s.io/jobset v0.3.0
```
Comment (kannon92) [inline go.mod:31]:
```suggestion
	sigs.k8s.io/jobset v0.3.1
```

**Comment by tenzen-y** [comment]
@kannon92 Hi! Are you still here?

I think we should include this in the next new minor version.
Since we will release the new minor version in the first week of February, could you complete this by its day?

**Comment by kannon92** [comment]
I updated the PR but I need someone with Kueue experience to look into these errors.

I see two groups of failures:

```
field.NotSupported(field.NewPath("spec", "tolerations").Index(2).Child("effect"), corev1.TaintEffect("not-valid"), nil),
Error is "cannot infer t" and these errors are in pkg/webhooks/resourceflavor_webhook_test.go and pkg/webhooks/workload_webhooks_test.go.
```

And the pkg/controller/job/pod has the following error while running make test.

```
--- FAIL: TestReconciler_ErrorFinalizingPod (0.00s)
    /home/kehannon/Work/kueue/pkg/controller/jobs/pod/reconciler.go:233: "level"=2 "msg"="Reconciling Job" "job"="ns/pod" "gvk"="/v1, Kind=Pod"
    /home/kehannon/Work/kueue/pkg/controller/jobs/pod/pod_controller_test.go:2213: Expected reconcile error (-want,+got):
          any(
        - 	e"connection refused: connection refused",
          )
    /home/kehannon/Work/kueue/pkg/controller/jobs/pod/reconciler.go:233: "level"=2 "msg"="Reconciling Job" "job"="ns/pod" "gvk"="/v1, Kind=Pod"
```

Happy to resolve them if someone else can suggest a fix. Otherwise, I'm not sure how much progress I can make as I haven't done a lot of Kueue development lately.

**Comment by kannon92** [comment]
Running `make test`:

```
=== Failed
=== FAIL: pkg/controller/jobs/pod TestReconciler_ErrorFinalizingPod (0.02s)
    reconciler.go:263: "level"=2 "msg"="Reconciling Job" "job"="ns/pod" "gvk"="/v1, Kind=Pod"
    pod_controller_test.go:2526: Expected reconcile error (-want,+got):
          any(
        -       e"connection refused: connection refused",
          )
    reconciler.go:263: "level"=2 "msg"="Reconciling Job" "job"="ns/pod" "gvk"="/v1, Kind=Pod"

=== Errors
pkg/webhooks/resourceflavor_webhook_test.go:112:23: cannot infer T (/home/kehannon/go/pkg/mod/k8s.io/apimachinery@v0.29.1/pkg/util/validation/field/errors.go:203:19)
pkg/webhooks/workload_webhook_test.go:263:23: cannot infer T (/home/kehannon/go/pkg/mod/k8s.io/apimachinery@v0.29.1/pkg/util/validation/field/errors.go:203:19)
pkg/webhooks/workload_webhook_test.go:349:23: cannot infer T (/home/kehannon/go/pkg/mod/k8s.io/apimachinery@v0.29.1/pkg/util/validation/field/errors.go:203:19)
```

@trasc At a quick glance you added
```
				field.NotSupported(field.NewPath("spec", "tolerations").Index(2).Child("effect"), corev1.TaintEffect("not-valid"), nil),
```
We are getting `cannot infer T` from usage of this in kube 1.29.

**Comment by kannon92** [comment]
> @kannon92 Hi! Are you still here?
> 
> I think we should include this in the next new minor version. Since we will release the new minor version in the first week of February, could you complete this by its day?

I don't think I'll be able to hit that february deadline as I'll be traveling.

**Comment by mimowo** [comment]
the first issue with cannot infer can probably be easily fixed by wrapping the nil with type `[]string(nil)`?

I'm not sure about the missing reconcile error. Probably can take a look tomorrow.

**Comment by kannon92** [comment]
> the first issue with cannot infer can probably be easily fixed by wrapping the nil with type `[]string(nil)`?
> 
> I'm not sure about the missing reconcile error. Probably can take a look tomorrow.

Thank you! That fix worked like a charm. Still got the reconcile error and it looks to also impact 1.29  e2e test

File: pkg/webhooks/resourceflavor_webhook_test.go [hunk 2]
```diff
@@ -109,7 +109,7 @@ func TestValidateResourceFlavor(t *testing.T) {
 					Value:    "v",
 					Effect:   corev1.TaintEffectNoSchedule,
 				}, ""),
-				field.NotSupported(field.NewPath("spec", "tolerations").Index(2).Child("effect"), corev1.TaintEffect("not-valid"), nil),
+				field.NotSupported(field.NewPath("spec", "tolerations").Index(2).Child("effect"), corev1.TaintEffect("not-valid"), []string{}),
```
Comment (trasc) [inline pkg/webhooks/resourceflavor_webhook_test.go:112]:
```suggestion
				field.NotSupported(field.NewPath("spec", "tolerations").Index(2).Child("effect"), corev1.TaintEffect("not-valid"), []corev1.TaintEffect{}),
```

**Comment by trasc** [comment]
For 
```
    pod_controller_test.go:2526: Expected reconcile error (-want,+got):
          any(
        -       e"connection refused: connection refused",
          )
    reconciler.go:263: "level"=2 "msg"="Reconciling Job" "job"="ns/pod" "gvk"="/v1, Kind=Pod"
```
It looks like the TypeMeta of the object in 

https://github.com/kubernetes-sigs/kueue/blob/b7fc93f5ba42ebbcd54345702be24c3f76d7ea26/pkg/controller/jobs/pod/pod_controller_test.go#L2492-L2507

is not populated, you could :

a. get the GVK from the clients schema 
```
                               kinds, _, err := client.Scheme().ObjectKinds(obj)
                               if err == nil {
                                       gvk = kinds[0]
                               }
```

b.  do a type assertion instead `_, isPod := obj.(*corev1.Pod)`

**Comment by mimowo** [comment]
> b. do a type assertion instead `_, isPod := obj.(*corev1.Pod)`

If this is just about unit tests, it is probably ok to go with b as simpler.

**Comment by tenzen-y** [comment]
> For the flaky e2e multikueue test we have #1659 under review.
> 
> For
> 
> ```
>     pod_controller_test.go:2526: Expected reconcile error (-want,+got):
>           any(
>         -       e"connection refused: connection refused",
>           )
>     reconciler.go:263: "level"=2 "msg"="Reconciling Job" "job"="ns/pod" "gvk"="/v1, Kind=Pod"
> ```
> 
> It looks like the TypeMeta of the object in
> 
> https://github.com/kubernetes-sigs/kueue/blob/b7fc93f5ba42ebbcd54345702be24c3f76d7ea26/pkg/controller/jobs/pod/pod_controller_test.go#L2492-L2507
> 
> is not populated, you could :
> 
> a. get the GVK from the clients schema
> 
> ```
>                                kinds, _, err := client.Scheme().ObjectKinds(obj)
>                                if err == nil {
>                                        gvk = kinds[0]
>                                }
> ```
> 
> b. do a type assertion instead `_, isPod := obj.(*corev1.Pod)`

@trasc I guess that the root cause of the TypeMeta not being populated is https://github.com/kubernetes-sigs/controller-runtime/pull/2633.

File: pkg/webhooks/resourceflavor_webhook_test.go (hunk 2 above)
Comment (kannon92) [reply pkg/webhooks/resourceflavor_webhook_test.go]: updated.

**Comment by kannon92** [comment]
Pushed a fix for each. Thank you for looking into that.

**Comment by kannon92** [comment]
/assign @trasc @tenzen-y
This is good to go now.

/retest

**Review by trasc — COMMENTED** [review]
/lgtm

**Review by tenzen-y — COMMENTED** [review]
/approve

**Pull request merged**